src/hooks/useReciever.ts

```diff
@@ -0,0 +1,123 @@
+  append: UseFieldArrayAppend<FormValues, 'reciever'>;
+  remove: UseFieldArrayRemove;
+};
+const validators = {
+  name: (value: string): string => {
+    return !value.trim() ? '이름을 입력해주세요.' : '';
+  },
+  phone: (value: string, allPhones: string[], currentIndex: number): string => {
+    if (!value.trim()) return '전화번호를 입력해주세요.';
+    const cleanPhone = value.replace(/-/g, '');
+    if (!/^01[016789][0-9]{3,4}[0-9]{4}$/.test(cleanPhone)) {
+      return '올바른 전화번호 형식이 아닙니다.';
+    }
+    const phones = allPhones.map((phone, idx) => (idx === currentIndex ? value : phone));
+    if (phones.filter((phone) => phone === value).length > 1) {
+      return '동일한 전화번호는 입력할 수 없습니다.';
+    }
+    return '';
+  },
+  count: (value: number): string => {
+    return !value || value < 1 ? '수량은 1개 이상이어야 합니다.' : '';
+  },
+};
+
+function useReciever({ open, onComplete, initialList, append, remove }: useRecieverProps) {
+  const [newList, setNewList] = useState<RecieverType[]>([]);
+  const [fieldErrors, setFieldErrors] = useState<FieldError[]>([]);
+  const handleChange = (idx: number, key: keyof RecieverType, value: string | number) => {
+    setNewList((prev) => prev.map((item, i) => (i === idx ? { ...item, [key]: value } : item)));
+    setFieldErrors((prev) => {
+      const next = [...prev];
+      if (!next[idx]) next[idx] = {};
+      let err = '';
+      const phones = newList.map((r, i) => (i === idx ? String(value) : r.phone));
+      switch (key) {
+        case 'name':
+          err = validators.name(String(value));
+          break;
+        case 'phone':
+          err = validators.phone(String(value), phones, idx);
+          break;
+        case 'count':
+          err = validators.count(Number(value));
+          break;
+        default:
+          next[idx] = { ...next[idx], [key]: err };
+      }
+      next[idx] = { ...next[idx], [key]: err };
+      return next;
+    });
+  };
+
+  const handleAdd = () => {
+    setNewList([...newList, { name: '', phone: '', count: 1 }]);
+    setFieldErrors([...fieldErrors, {}]);
+  };
+
+  const handleRemove = (idx: number) => {
+    setNewList((prev) => prev.filter((_, i) => i !== idx));
+    setFieldErrors((prev) => prev.filter((_, i) => i !== idx));
+  };
+
+  const handleComplete = () => {
+    const errors: FieldError[] = [];
+    let hasError = false;
+    const phones = newList.map((r) => r.phone);
+    newList.forEach((item, idx) => {
+      const err: FieldError = {};
+      err.name = validators.name(item.name.trim());
+      err.phone = validators.phone(item.phone.trim(), phones, idx);
+      err.count = validators.count(item.count);
+      if (err.name !== '' || err.phone !== '' || err.count !== '') hasError = true;
+      errors[idx] = err;
+    });
+
+    setFieldErrors(errors);
+    if (hasError) return;
+
+    for (let i = initialList.length - 1; i >= 0; i--) {
+      remove(i);
+    }
+
+    newList.forEach((item) => {
+      append(item);
+    });
+    onComplete(newList);
+  };
```

Comment:
handleComplete 함수에서 처리하는 로직이 많아지면, 처음 코드를 접하는 사람 입장에서 함수의 역할을 직관적으로 이해하기 어려워질 수 있습니다.

현재 이 함수는 크게 두 가지 역할을 수행하고 있습니다:
- 유효성 검사 (에러를 찾는다)
- 상태 갱신 및 완료 처리 (정보를 업데이트한다)

각 역할을 함수로 분리하면 코드의 목적이 더 분명해지고, 유지보수와 테스트에도 유리합니다. 예를 들어 아래와 같이 리팩토링해볼 수 있을거 같아요!

```suggestion
  const handleComplete = () => {
  const errors = validateNewList(newList);
  setFieldErrors(errors);

  if (errors.some((err) => err.name || err.phone || err.count)) return;

  updateList();
  onComplete(newList);
};

const validateNewList = (list: RecieverType[]): FieldError[] => {
  const phones = list.map((r) => r.phone);
  return list.map((item, idx) => ({
    name: validators.name(item.name.trim()),
    phone: validators.phone(item.phone.trim(), phones, idx),
    count: validators.count(item.count),
  }));
};

const updateList = () => {
  for (let i = initialList.length - 1; i >= 0; i--) {
    remove(i);
  }
  newList.forEach((item) => {
    append(item);
  });
};
```